Post-submission clean-up

almaeder opened the pull request:
This addresses major points of vincent-maillou/quatrex#56:

- Clean up `qttools.__init__`
- Communication backend
- `serinv` backend clean-up
- One general structure for dsbsparse and dsdbspars (only dsdbsparse)

vetschn committed:
Added USE_NCCL env variable

alexnick83 committed:
Added free/allocate data methods to the datastructures.

vetschn committed:
Switch to xp.empty and fix a GPU bug

almaeder committed:
Small fixes

vetschn committed:
Refactor memory management in DSBSparse and DSDBSparse to use free_mempool function

almaeder committed:
Memory improvements

almaeder committed:
reordering for alps/frontier

almaeder committed:
Refactor communication type handling in DSBSparse and DSDBSparse; replace USE_NCCL with ALLTOALL_COMM_TYPE and update barrier calls to use global_comm.

vetschn committed:
MPI block comm

almaeder committed:
Custom reduce, remove sum, remove nccl comms creation

vetschn committed:
Factor truthy environment variable check into separate method

Also formatting, and linter chekcs

vetschn committed:
First draft of backend-agnostic communication interface

almaeder committed:
Implemented tests for the comm. backend

almaeder committed:
Blunder :skull:

vetschn committed:
Fix _nccl_comm attribute name

almaeder committed:
all gather v

vetschn committed:
New buffer aliasing checks and backend override.

almaeder committed:
Moved allgatherv

vetschn committed:
Override functionality :+1:

vetschn committed:
Refactor tests to use new communication framework and replace host_xp with numpy

- Updated test files in `tests/datastructures`, `tests/greens_function_solver`, `tests/kernels`, `tests/rgf_dist`, and `tests/utils` to replace instances of `host_xp` with `numpy`.
- Introduced a `setup_module` function in several test files to configure the communication settings using the new `comm` singleton.
- Adjusted block communication references from `block_comm` to `comm.block` for consistency with the new communication framework.
- Ensured that all tests are compatible with the updated communication methods and numpy arrays.

almaeder committed:
Moved raw /jit kernels
Renamed kernels

almaeder committed:
Ammend import path, zero_like gpu issue,

almaeder committed:
Renamed cuda to cupy in kernels datastructure

almaeder committed:
Refactor find_inds functionality and update densify_block test to support kernel usage

vetschn committed:
Remove many prints

vetschn committed:
:bug: fixed

vetschn committed:
Add aliases for the non-distributed case

vetschn committed:
Negative block-indexing is not supported by the distributed datastructure

vetschn committed:
Add a warning for the aliases.

almaeder committed:
Refactor DSBSparse to DSDBSparse

- Updated all references from DSBSparse to DSDBSparse across multiple files to maintain consistency after renaming.
- Removed deprecated densify_selected_blocks function from sparse_utils.py.
- Ensured all documentation and type hints reflect the changes from DSBSparse to DSDBSparse.

almaeder committed:
Renaming and bug fixes

-Renamed kernels from dsb to dsdb
-Removed product_sparsity_pattern_dsbsparse
- Fixed bug in _check_bufs_consistent
- Fixed bug in dsdbcoo._check_commensurable

almaeder committed:
Refactor and enhance tests for DSDBSparse and related structures

- Updated imports to reflect changes in data structure naming from DSBCOO, DSBCSR, and DSBSparse to DSDBCOO, DSDBCSR, and DSDBSparse.
- Modified test files to use the new data structure names, ensuring consistency across tests.
- Enhanced the configuration of communication settings in distributed tests to support different environments (Cupy and Numpy).
- Removed deprecated tests related to DSBSparse and adjusted the product sparsity pattern tests to focus on DSDBSparse.
- Cleaned up unnecessary comments and code in utility functions to improve readability and maintainability.

almaeder committed:
- Fixed product_sparsity_dsdbsparse tests with multiple block comm ranks
- Refactor DSDBSparse type hints
- Update test configurations for block communicator sizes

almaeder committed:
Enhance DSDBCOO and related tests for symmetry handling and distribution logic

- fixed get item nnz
- test for get item nnz

almaeder committed:
Fixed stucked tests by always having a non zero diagonal

almaeder committed:
Refactor DSDBCOO and DSDBSparse classes

- Changed handling of padding such that data is always contiguous
- Adapted dtranspose, fill_diagonal, and data setter/getter for this
- Cleaned DSDCOO _set_item (able to use .data since advanced indexing is not used anymore)

almaeder committed:
Refactor tests to support better the GPU tests and improve symmetry handling

- Updated test configurations to differentiate between GPU and CPU environments, ensuring proper MPI settings based on the execution context (cupy vs numpy).
- Added checks to skip tests when GPU-aware MPI is not available and the block communicator lacks an NCCL communicator due to missing host mpi implementation.
- Refactored tests in `test_dsdbsparse_dist.py` to utilize a new helper function for creating COO matrices with symmetry considerations.
- Enhanced tests in `test_routines.py`, `test_selected_inv.py`, `test_rgf_dist.py`, and others to ensure compatibility with the new MPI configurations.
- Fused certain distributed and non-distributed tests to reduce code dublication. The tests need to be run now with --with-mpi such that tests with no number of ranks requirements are also run.

almaeder committed:
Precomputation of non-zero rows and columns for nnz dist, and improve product sparsity pattern test with periodic boundary conditions. Test failed before the change.

vetschn commented on src/qttools/__init__.py:
```diff
@@ -1,184 +1,72 @@
 
-        host_xp = xp
-        pinned_xp = None
+def strtobool(s: str, default: bool | None = None) -> bool:
```
Just realized this function is not as practical as it could be, should also check for `s is None` and return default in that case

vetschn commented on src/qttools/comm/comm.py:
```diff
@@ -0,0 +1,549 @@
+def _check_bufs_aliased(sendbuf: NDArray, recvbuf: NDArray) -> bool:
+    """Checks if the send and receive buffers are aliased.
+
+    This is done by checking if the memory addresses of the two buffers
+    are the same.
```
Lies

alexnick83 replied:
Maybe say that this returns True if the memory addresses match or the buffers overlap.

alexnick83 committed:
Fixed boundary conditions for spillover tests.

vetschn committed:
Remove unnecessary BLOCK_COMM_SIZE environment variable from MPI test execution

vetschn reviewed (commented):
Left some comments, the `test_product_sparsity_dsdbsparse_spillover` tests are still failing

almaeder commented on tests/comm/test_comm.py:
```diff
@@ -0,0 +1,306 @@
+
+
+@pytest.fixture(scope="function")
+def reset_comm():
+    # Setup: set all members of the singleton comm to None
+    for attr in comm.__dict__:
+        setattr(comm, attr, None)
+    comm._is_configured = False
+
+    # Yield control to the test function
+    yield
```
👍 This is from before `override`

alexnick83 committed:
Fixed boundary conditions for spillover tests.

alexnick83 committed:
Merge branch 'use-nccl-flag' of github.com:vincent-maillou/qttools into use-nccl-flag

alexnick83 committed:
Fixed boundary conditions for spillover tests.

vetschn committed:
:bug: fixed (typo)

vetschn committed:
Remove warnings for using 'blocks' and 'sparse_blocks' with domain-distribution

vetschn committed:
Remove symmetry checks and improve error handling for deallocated data.

vetschn committed:
Refactor DSDBSparse access methods to always use 'blocks' and never 'local_blocks'

- Removed the 'local_blocks' and 'sparse_local_blocks' properties from the DSDBSparse class.
- Updated all references to use 'blocks' instead of 'local_blocks'.

almaeder committed:
:bug:

vetschn committed:
Formatting

vetschn committed:
Explicitely cast to int32.

This was causing problems with the RawKernel backend.

vetschn committed:
Rank 0 will let us know during .load(path) if the file actually exists.

vetschn committed:
Make sure the test data is what the kernel expects.

almaeder committed:
Use override in test comm

almaeder committed:
Refactor strtobool calls to remove default string values for environment variables

almaeder commented on src/qttools/datastructures/dsdbsparse.py:
```diff
@@ -101,24 +163,29 @@ def __init__(
-            local_data.shape[-1], stack_comm.size, strategy="greedy"
+            data.shape[-1], comm.stack.size, strategy="greedy"
+        )
+        self.nnz_section_sizes_host = nnz_section_sizes
+        self.nnz_section_offsets_host = np.hstack(
+            ([0], np.cumsum(self.nnz_section_sizes_host))
```
Only to avoid the memory copy when doing int(), the DSDBSparse __init__ needs further refactoring.

almaeder committed:
Refactor documentation to clarify buffer handling and remove unused communicator references

vetschn committed:
Put the FileNotFoundError back

vetschn committed:
Add a small note in the tests for the distributed load.

The tests only work if all ranks have access to the same file system.

almaeder committed:
Refactor data slicing in DSDBSparse class for improved clarity

almaeder committed:
:bug: in data_slice...

alexnick83 commented on src/qttools/__init__.py:
```diff
@@ -1,184 +1,73 @@
-            host_xp = xp
-            pinned_xp = None
+# Suppress warnings from the jit module if not rank 0.
+if global_comm.rank != 0:
```
I guess this runs before the `comm` object is created

alexnick83 commented on src/qttools/comm/comm.py:
```diff
@@ -0,0 +1,557 @@
+        return xp.take(recvbuf, indices, axis=axis)
+
+
+class QuatrexCommunicator:
```
Shouldn't this also have, except `barrier`, also global-comm methods for bcast, reduce, etc.?

vetschn replied:
Would be neat :+1: Should that also support different communication backends in your opinion?

alexnick83 replied:
Good question. I think this is always the MPI comm now, so let's keep it like that but consider a TODO for the future?

vetschn committed:
Fix the spillover test by making the data periodic.

This is the same that is done for the domain-distributed case.

alexnick83 commented on src/qttools/datastructures/dsdbcsr.py:
```diff
@@ -23,14 +23,19 @@ def _upper_triangle(rows: NDArray, cols: NDArray) -> tuple[NDArray, NDArray, NDA
-class DSBCSR(DSBSparse):
-    """Distributed stack of block-compressed sparse row matrices.
+class DSDBCSR(DSDBSparse):
+    """A Distributed Stack of Distributed Block-accessible CSR matrices.
```
I would just say a distributed stack of block-distributed ... no need to change the acronym though.

vetschn replied:
Opened an issue here: #155

alexnick83 commented on src/qttools/datastructures/dsdbsparse.py:
```diff
@@ -644,120 +695,29 @@ def _dtranspose(
-            stack_comm.Alltoall(MPI.IN_PLACE, _data_host)
-            self._data = xp.array(_data_host)
+        receive_buffer = xp.empty_like(self._data)
+        comm.stack.all_to_all(self._data, receive_buffer)
```
Do we want to keep the timer here?

vetschn replied:
New issue here: https://github.com/vincent-maillou/quatrex/issues/61

alexnick83 commented on src/qttools/greens_function_solver/rgf.py:
```diff
@@ -297,14 +297,14 @@ def selected_solve(
 
             # We need to write the last diagonal blocks to the output.
-            xl_.blocks[-1, -1] = 0.5 * (
+            xl_.blocks[j, j] = 0.5 * (
```
Why this change? In principle, I would avoid using Python's (lack of) scoping rules.

vetschn replied:
The `.block` attribute does not support negative indices, but yes, this is sketchy. Instead in the `BlockIndexer` we could detect whether we are in distributed mode or non-distributed mode and allow/disallow negative indices based on that? Or we use `xl.num_blocks - 1` as index here.

alexnick83 replied:
I see. I would go with `num_blocks - 1`, if it is block-distribution agnostic.

vetschn committed:
Less abuse of Python scoping "rules"

vetschn committed:
Expose RGFDist in the init
